panther_analysis_tool/testing.py

```diff
@@ -0,0 +1,254 @@
+    # to response attributes by API & FE.
+    # TODO: provide a field name translation step if necessary
+    detectionId: Optional[str]  # pylint: disable=C0103
+    genericError: Optional[str]  # pylint: disable=C0103
+    # TODO:
+    error: Optional[TestError]
+    errored: bool
+    passed: bool
+    trigger_alert: Optional[bool]
+    functions: TestResultsPerFunction
+
+
+@dataclass
+class TestExpectations:
+    """Contains the expected values for performing assertions"""
+
+    detection: bool
+    # TODO: include assertions for remaining functions, e.g title and alert context
+
+
+@dataclass
+class TestSpecification:
+    """The structure of a test case"""
+
+    id: str  # pylint: disable=C0103
+    name: str
+    data: Dict[str, Any]
+    mocks: List[Dict[str, Any]]
+    expectations: TestExpectations
+
+
+# pylint: disable=too-few-public-methods
+class TestCaseEvaluator:
+    """Translates detection execution results to test case results,
+    by performing assertions and determining the status"""
+
+    def __init__(self, spec: TestSpecification, detection_result: DetectionResult):
+        self._spec = spec
+        self._detection_result = detection_result
+
+    def _get_result_status(self) -> bool:
+        """Get the test status - passing/failing"""
+
+        # Title/dedup functions are executed unconditionally
+        # (regardless if the detection trigger_alert or not) during testing.
+        if self._spec.expectations.detection == self._detection_result.detection_output:
+            # Any error should mark the test as failing
+            return not self._detection_result.errored
+        # expectations didn't match for the detection output
+        return False
+
+    def _get_generic_error_details(self) -> Tuple[Optional[Exception], Optional[str]]:
+        generic_error = None
+        generic_error_title = None
+        if self._detection_result.input_exception is not None:
+            generic_error = self._detection_result.input_exception
+            generic_error_title = "Invalid event"
+        elif self._detection_result.setup_exception is not None:
+            generic_error = self._detection_result.setup_exception
+        return generic_error, generic_error_title
+
+    def interpret(self) -> TestResult:
+        """Evaluate the detection result taking into account
+        the errors raised during evaluation and
+        the test specification expectations"""
+        function_results = dict(
+            detectionFunction=FunctionTestResult.new(
+                self._detection_result.detection_output,
+                self._detection_result.detection_exception,
+            )
+        )
+
+        # We don't include output from other functions
+        # unless the test was expected to match and trigger an alert.
+        # If the test fails, providing all the output provides a faster feedback loop,
+        # on possible additional failures.
+        if self._detection_result.trigger_alert or (
```

Comment:
Alright, now I think I get it 😄 . So my suggestion would be here to change the condition to check for the configured alert value of each detection type. Something like:

```python
  def _get_detection_alert_value(self):
    if self._detection_result.detection_type == TYPE_POLICY:
      return Policy.matcher_alert_value
   elsif self._detection_result.detection_type == TYPE_RULE:
      return Rule.matcher_alert_value
```

```python
if self._spec.expectations.detection == self._get_detection_alert_value()
```

Doesn't this cover all cases? If the test is expected to trigger an alert we should always display the aux functions results, regardless of whether it actually triggered the alert or not.